### PR TITLE
feat(stdlibs): Fuzz for String

### PR DESCRIPTION
**Pull request by notJoon**
## Description

I have implemented basic Fuzz functions in the testing package and a function to apply fuzzing to string data.

To transform string data, I used genetic algorithms. Additionally, I implemented a Uniform distribution to generate random values and then used the central limit theorem to approximate a normal distribution, thereby enabling the creation of random booleans in the random file.

In this PR, only the string type is handled. I plan to support other types in the future.

### CODE REVIEW

File: gnovm/stdlibs/testing/fuzz.gno [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+		return InsertDelete(sf.Value, 0.1)
+	} else {
+		rs := []rune(sf.Value)
+		if len(rs) > 0 {
+			index := randRange(0, len(rs)-1)
+			rs[index] = randomASCIIChar()
+		}
+
+		return string(rs)
+	}
```
Comment (zivkovicmilos) [inline gnovm/stdlibs/testing/fuzz.gno:49]: You can simplify this as:
```go
if GenerateRandomBool(0.2) {}
    return InsertDelete(sf.Value, 0.1)
}

rs := []rune(sf.Value)
if len(rs) > 0 {
    index := randRange(0, len(rs)-1)
    rs[index] = randomASCIIChar()
}

return string(rs)
```

File: gnovm/stdlibs/testing/fuzz.gno [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+func (ind *Individual) calculateFitness() {
+	ind.Fitness = len(ind.Fuzzer.String())
+}
+
+// Selection selects individuals from the population based on their fitness.
+//
+// Use roulette wheel selection to select individuals from the population.
+// ref: https://en.wikipedia.org/wiki/Fitness_proportionate_selection
+// TODO: resolve scaling issues with large fitness values.
```
Comment (zivkovicmilos) [inline gnovm/stdlibs/testing/fuzz.gno:75]: Leftover?

File: gnovm/stdlibs/testing/fuzz.gno [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+		for i := range args {
+			args[i] = input
+		}
+
+		run(nil, args...)
+	}
+}
+
+// Add adds test values to initialize the corpus.
+// TODO: length assertion. if the input length exceeds or is less than the test target function, it should be fatal.
```
Comment (zivkovicmilos) [inline gnovm/stdlibs/testing/fuzz.gno:174]: Leftover?

File: gnovm/stdlibs/testing/fuzz.gno [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+	for gen := 0; gen < generations; gen++ {
+		population = Selection(population)
+		newPopulation := make([]*Individual, 0, len(population))
+		for i := 0; i < len(population); i += 2 {
+			if i+1 < len(population) {
+				child1, child2 := Crossover(population[i], population[i+1])
+				newPopulation = append(newPopulation, child1, child2)
+			} else {
+				newPopulation = append(newPopulation, population[i])
+			}
```
Comment (zivkovicmilos) [inline gnovm/stdlibs/testing/fuzz.gno:209]: You can simplify this as:
```go
if i+1 < len(population) {
    child1, child2 := Crossover(population[i], population[i+1])
    newPopulation = append(newPopulation, child1, child2)

    continue
}

newPopulation = append(newPopulation, population[i])
```

File: gnovm/stdlibs/testing/random.gno [hunk 5]
```diff
@@ -0,0 +1,75 @@
+package testing
+
+import (
+	"math"
+	"time"
+)
+
+// Internal state for the random number generator.
+var x uint64 = 42
```
Comment (zivkovicmilos) [inline gnovm/stdlibs/testing/random.gno:9]: What do you think about moving this global into a structure instead?
Having it global and named `x` can be super confusing, not to mention its usage can't be parallelized

File: gnovm/stdlibs/testing/random_test.gno [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	"strconv"
+	"time"
+)
+
+func updateSeed() {
+	seed := time.Now().UnixNano()
+	_srand(seed)
+}
+
+// TODO: use chi-squared test to check number distribution follows uniform distribution.
```
Comment (zivkovicmilos) [inline gnovm/stdlibs/testing/random_test.gno:14]: Leftover?

File: gnovm/stdlibs/testing/random_test.gno [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+		}
+	}
+
+	lvm := len(valueMap)
+	if lvm > maxIter || lvm == 0 {
+		t.Errorf("len(valueMap) = %v, want <= %v", lvm, maxIter)
+	}
+}
+
+// TODO: use chi-squared test to check number distribution follows normal distribution.
```
Comment (zivkovicmilos) [inline gnovm/stdlibs/testing/random_test.gno:39]: Leftover?

**Review by zivkovicmilos — APPROVED** [review]
Looks good 💯 

Please check some comments I've left, otherwise it should be good to go

File: gnovm/stdlibs/testing/fuzz.gno [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+	index := randRange(0, len(runes)-1)
+	runes[index] = randomASCIIChar()
+	return &StringFuzzer{Value: string(runes)}
+}
+
+func (sf *StringFuzzer) InsertDeleteMutate(p float64) Fuzzer {
+	return &StringFuzzer{Value: InsertDelete(sf.Value, p)}
+}
+
+func (sf *StringFuzzer) Fuzz() string {
```
Comment (petar-dambovaliev) [inline gnovm/stdlibs/testing/fuzz.gno:38]: You could use less scopes so the code is left leaning and needs less brain power to read.

**Review by petar-dambovaliev — APPROVED** [review]

File: gnovm/stdlibs/testing/random_test.gno (hunk 7 above)
Comment (notJoon) [reply gnovm/stdlibs/testing/random_test.gno:39]: It's necessary for a more accurate test. but it also seems there's no need to implement anything extra just for the test. so yes leftover, I think so.

File: gnovm/stdlibs/testing/random.gno (hunk 5 above)
Comment (notJoon) [reply gnovm/stdlibs/testing/random.gno:9]: I agree with your suggestion. However, I'm still tryin to decide whether to use a formula-based function (current implementation) or use chacha20's function (which already implemented in gno libs) to generate random numbers. So, I will modify it later if I decided. 🤔

**Comment by zivkovicmilos** [comment]
@notJoon 
The `_test.cmd` gnovm make directive is failing (tests). Please see what this is about before we merge 🙏

**Comment by notJoon** [comment]
> @notJoon The `_test.cmd` gnovm make directive is failing (tests). Please see what this is about before we merge 🙏

Resolved it by using type assertion instead of using `ufmt`. thanks!

**Pull request merged**